### ISSUE TITLE
Using dynamic classes

### ISSUE THREAD

**Issue by TedMeftah** [windicss/svelte-windicss-preprocess#117]
After finding out that using `windi` from `windicss/helpers` did not work in dev mode (windicss/windicss#244). I noticed that in the code of this package there is a check for the for the `windi` template:
https://github.com/windicss/svelte-windicss-preprocess/blob/2c7003c71b18c69d88b9ff14ff1f1a13678b9220/src/index.ts#L133-L136
so we don't actually need to import the `windi` helper we can just fake it like this:
```js
function windi(parts, ...vars) {
  return parts.reduce((out, part, i) => (out += part + (vars[i] || '')), '')
}
```
which make this work
```svelte
<div class={`bg-red-50`}>This does not work since it's an expression</div> 

<div class={windi`bg-red-100`}>This works</div>
```
of course this is not so useful, since the helper does update the classes in runtime, so all these examples below don't work:

```svelte
<script>
  function windi(parts, ...vars) {
    return parts.reduce((out, part, i) => (out += part + (vars[i] || '')), '')
  }

  let shade = 300
  let dynamicClass = windi`bg-red-${shade}`
  let toggle = false
</script>

<button on:click={() => toggle = !toggle}>Toggle [{toggle}]</button>

<div class={windi`bg-red-${shade}`}>My background is dynamic</div>

<div class={dynamicClass}>My background is dynamic</div>

<div class={windi`${toggle ? 'bg-blue-500' : 'bg-red-500'}`}>My background is togglable</div>
```
I feel that the last example should work but it does not (```class={windi`${toggle ? 'bg-blue-500' : 'bg-red-500'}`}```).

what's weird is when I do this:
```svelte
<div class={toggle ? windi`bg-yellow-500` : windi`bg-blue-500`}>My background is togglable</div>
```
it partially works, only taking the first `windi` expression (windi`bg-yellow-500`) but it toggle just fine.

I found the best way to do dynamic styles is to use custom classes and `@apply`, 

```svelte
<script>
  let toggle = true
</script>

<button on:click={() => toggle = !toggle}>Toggle [{toggle}]</button>
<div class={toggle ? "red" : "blue"}>My background is togglable</div>

<style>
  .red {
    @apply bg-red-800;
  }
  .blue {
    @apply bg-blue-800;
  }
</style>
```
I think we should advocate this way to use dynamic classes in the docs, since this avoids adding code to the runtime.

An improvement that can be added to this is allow the mixing of windicss classes, because this does not work right now: 

```svelte
<script>
  function windi(parts, ...vars) {
    return parts.reduce((out, part, i) => (out += part + (vars[i] || '')), '')
  }
  
  let toggle = true
</script>

<button on:click={() => toggle = !toggle}>Toggle [{toggle}]</button>

<div class={toggle ? "red text-4xl" : "blue text-6xl"}>My background is togglable</div>

<div class={toggle ? windi`red text-4xl` : windi`blue text-6xl`}>My background is togglable</div>

<style>
  .red {
    @apply bg-red-800;
  }
  .blue {
    @apply bg-blue-800;
  }
</style>
```
it only render the `red` and `blue` classes, and like before only render `windi`red text-4xl` in the second example.

**Comment by alexanderniebuhr** [comment]
TLDR: at all I think you are using `windi``` to often

for dynamic classes on variable, the svelte way would be
```svelte
<div class:bg-red-800={toggle} class:bg-blue-800={toggle}>My background is togglable</div>
```
---
> so we don't actually need to import the `windi` helper we can just fake it like this

you need to import it, since `windi``` is used to interpret classes in runtime, the code in the pre-processor just makes sure it does add that classes to safelist

---

> ```svelte
> <div class={`bg-red-50`}>This does not work since it's an expression</div> 
> ```

that should work, maybe use `' or "`

---



> ```svelte
> <div class={windi`${toggle ? 'bg-blue-500' : 'bg-red-500'}`}>My background is togglable</div>
> ```

just use
```svelte
<div class={toggle ? 'bg-blue-500' : 'bg-red-500'}>My background is togglable</div>
```

---

> ```svelte
> <div class={toggle ? "red text-4xl" : "blue text-6xl"}>My background is togglable</div>
> ```

we would need `debug + verbosity 3, 4 & 5` for this bug

**Comment by TedMeftah** [comment]
Thank you @alexanderniebuhr, I agree that's too much use for `windi` but they were just examples to test all the possible cases.

I could not import `windi` since it has an issue while running `yarn dev` (windicss/windicss#244). and since I want to avoid adding code to the runtime I will try to avoid it.

I did more dinging in the code and I noticed that this does not work:

```svelte
<div class={toggle ? windi`bg-yellow-500` : windi`bg-blue-500`}>My background is togglable</div>
```
because it seems that we are only taking the first match of `windi` expression per line:
https://github.com/windicss/svelte-windicss-preprocess/blob/2c7003c71b18c69d88b9ff14ff1f1a13678b9220/src/index.ts#L136

and for this:
```svelte
<div class={toggle ? "red text-4xl" : "blue text-6xl"}>My background is togglable</div>
```

it looks like we do format the code first, then we only match classes like `class="..."` or `class='...'` since `'` get converted to `"` when formatted using prettier.

This ignores ```class=`...` ``` and `class={"..."}` which are valid classes and should be checked too.

This is the current regex:
```
[DEBUG] regex parse matcher ((class|w:sm|w:-sm|w:+sm|w:md|w:-md|w:+md|w:lg|w:-lg|w:+lg|w:xl|w:-xl|w:+xl|w:hover|w:focus|w:visited|w:checked)=["])([^"]*)(["])
```

**Comment by alexanderniebuhr** [comment]
> ```svelte
> <div class={toggle ? windi`bg-yellow-500` : windi`bg-blue-500`}>My background is togglable</div>
> ```

I don't think this would be good way of writing this, so not sure if we make it work. I think to use windi`` i would suggest doing
```svelte
<script>
let myDynamicToggleClasses = windi`bg-yellow-500`
</script>
<div class={toggle ? myDynamicToogleClassses}>My background is togglable</div>
```

---

> ```svelte
> <div class={toggle ? "red text-4xl" : "blue text-6xl"}>My background is togglable</div
> ```

yeah i will fix this right now.. seems that there are some bugs

**Comment by alexanderniebuhr** [comment]
@TedMeftah do not try to understand the codebase now.. check branch `next`
If you would like to help, I think the best currently (as you did above, is give me a list of test cases)

**Comment by alexanderniebuhr** [comment]
@TedMeftah you said, you would like to contribute. if you have time, feel free to open PR onto the `next` branch, where you add `.svelte` files with test cases in this directory. filename should be descriptive and it should be one test case per file only.
https://github.com/windicss/svelte-windicss-preprocess/tree/next/tests/assets/input

**Comment by alexanderniebuhr** [comment]
still tracking by [v4.1](https://github.com/windicss/svelte-windicss-preprocess/issues/118)

**Comment by alexanderniebuhr** [comment]
should be fixed with 7482f9a1f58dc003c1484eb80cc07e5d49d5b476
